cpp/src/fil/internal.cuh

```diff
@@ -266,6 +302,90 @@ struct forest_params_t {
 /// FIL_TPB is the number of threads per block to use with FIL kernels
 const int FIL_TPB = 256;
 
+static const uint32_t max_precise_int_float = 1 << 24;  // 16'777'216
+
+struct cat_feature {
+  int max_matching = -1, n_nodes = 0;
+  void reduce_with(cat_feature b)
+  {
+    max_matching = std::max(max_matching, b.max_matching);
+    n_nodes += b.n_nodes;
+  }
+};
+typedef std::vector<cat_feature> v_cat_feature;
+struct categorical_branches {
+  // arrays from each node ID are concatenated first, then from all categories
+  uint8_t* bits = nullptr;
+  // largest matching category in the model, per feature ID
+  int* max_matching = nullptr;
+  size_t bits_size = 0, max_matching_size = 0;
+
+  // set count is due to tree_idx + node_within_tree_idx are both ints, hence uint32_t result
+  template <typename node_t>
+  __host__ __device__ __forceinline__ int get_child(const node_t& node,
```

Comment:
This structure is responsible for bitmaps only, and only needs to return whether for a given bitmap the bit is set. It doesn't need to know anything about the tree nodes, or what to do if the node is not categorical.